### ISSUE TITLE
Option to SAVE by Default, with AUTOMATIC and RECURSIVE Support

### ISSUE THREAD

**Issue by jcburley** [CodethinkLabs/ofc#23]
Much of our codebase is compiled with an option making `SAVE` the default (for local variables/arrays at least; not sure about the affect on COMMON blocks, though IME that really makes no difference in code generation).

So, for (implicitly) RECURSIVE procedures, we use the AUTOMATIC statement, akin to SAVE but having the opposite effect.

This yields:

```
Use of AUTOMATIC keyword is non-standard and deprecated, ignoring
```

While it's possible all the compilers we'll use on ofc's output will support the same option to make `SAVE` default, ideally ofc would offer a similar option and convert the code by eliminating AUTOMATIC (as it now apparently does) and inserting SAVE statements for whatever is appropriate -- which, superficially, would include all local variables/arrays not declared AUTOMATIC, but could be smart enough (using flow analysis) to avoid simple cases where it's obviously not needed.

If that's too difficult (for now), at least please do optionally preserve AUTOMATIC in the output, or the resulting code likely won't work when recursion actually happens at run time. (This might also deserve to be the default.)

Further, use of any AUTOMATIC keyword implies the programmer either intended the possibility of recursion or worried that a sufficiently large amount of storage would unduly increase the size of the executable (and VM) even when the procedure wasn't active -- yet wasn't worried about stack.

Let's go with the former case and (perhaps optionally) support adding the RECURSIVE keyword on all such procedures (maybe optionally on all procedures, perhaps unless a given procedure can be shown to not possibly need it).

**Comment by ben-brewer-codethink** [comment]
The main problem with outputting AUTOMATIC is that it's not in Fortran 90 and it's not supported by modern compilers. If we do this then gfortran cannot be run over the code, so currently we can't even test it.
STATIC and AUTOMATIC are completely deprecated and only STATIC has a modern alternative because AUTOMATIC is the assumed default in most cases.
I can easily add an option for printing it, because we used to do that, but it'll only be useful for legacy tools anyhow.

I think your point about the RECURSIVE keyword is important here. The RECURSIVE keyword is effectively the modern alternative to AUTOMATIC, so if we're not outputting that then our output is incorrect yeah.

**Comment by ben-brewer-codethink** [comment]
Hmm, maybe I'm misinterpreting this.
I had assumed that AUTOMATIC wasn't supported in Fortran 90 because the following won't compile:
INTEGER, AUTOMATIC :: A
But I guess the AUTOMATIC statement itself isn't deprecated, so we could probably handle it separately like EXTERNAL and friends.

**Comment by jcburley** [comment]
I don't think AUTOMATIC has ever been in any Fortran standard.

Also, by STATIC, do you perhaps mean SAVE? C uses `static` for what Fortran uses `SAVE` (mostly; Fortran overloads `SAVE` for COMMON areas to mean something vaguely similar, while C overloads `static` for...pretty much anything one can think of!).

Anyway, what I'm looking for, simplified, is to have code like this:

```
! Compile me with -fno-automatic (the default is -fautomatic):
SUBROUTINE X
REAL Y
INTEGER Z
AUTOMATIC Z
!...might call X again; want a distinct copy of Z, but same Y
END
```

Converted into:

```
RECURSIVE SUBROUTINE X
REAL, SAVE :: Y  ! or REAL Y and SAVE Y as separate statements
INTEGER Z
...
END
```

Also note that, IIRC, RECURSIVE by itself doesn't imply anything about SAVE versus (the non-standard) AUTOMATIC. All it does is tell the compiler that the procedure might call itself, directly or indirectly; otherwise, the compiler is free to assume it won't do that, which can enable certain optimizations, including (if this is actually helpful) putting things in static memory instead of on the stack. Even if it did that sort of thing (in the absence of RECURSIVE), SAVE still tells it something important: unless it can prove unnecessary, it should commit (write back) values of SAVEd variables before returning to the caller on the grounds that they might be read by the next invocation. So, without e.g. `SAVE I`, the compiler would be free to drop the assignment in something like `I = 3; RETURN`. (That's true regardless of RECURSIVE.)

**Comment by ben-brewer-codethink** [comment]
I've added the "--print-automatic" option to allow reprinting of the AUTOMATIC attribute in the output. It's disabled by default as gfortran doesn't support it at all.
Commit for this fix is: a8ce21fcce86543c5670bd715c57e02f827af1d2

I'm leaving this issue open until we add RECURSIVE to function declarations containing AUTOMATIC variables.

**Comment by ben-brewer-codethink** [comment]
I've implemented printing the RECURSIVE keyword on FUNCTION and SUBROUTINE scopes where the body contains one or more AUTOMATIC variables. Note that when the "--print-automatic" option is used we don't print RECURSIVE and simply re-print the automatics instead to be as close to the original code as possible.
